### ISSUE TITLE
Add Litepcie bridge support to wishbone-tool

### ISSUE THREAD

**Issue by dkozel** [litex-hub/wishbone-utils#25]
I'm still very much figuring out how all these pieces connect together, but I believe that the wishbone-tool would be useful for interacting with a LiteX SoC connected to a host via LitePCIe.

Initially I'm just looking for ways to peek and poke registers defined in the CSR table. In C there is an example in the litepcie_util tool.
https://github.com/enjoy-digital/netv2/blob/master/software/user/litepcie_util.c#L436

But I think the larger wishbone server functionality is useful? and needs a different set of calls.
https://github.com/enjoy-digital/netv2/blob/master/software/user/litepcie_util.c#L447

Please do let me know if I've taken a wrong turn in my understanding.

**Comment by xobs** [comment]
Your understanding is mostly correct.  `wishbone-tool` could be used to do this, I just don't have access to any of the following:

1. A Linux system, with
2. A device connected via LitePCIe

I gather PCIe is one of those not-really-cross-platform things that we'd initially only support on Linux, similar to how we only support SPI on ARM Linux: https://github.com/litex-hub/wishbone-utils/blob/master/wishbone-tool/src/bridge/spi/mod.rs#L53-L61

As for *how* to actually access PCIe, on Linux it seems to be frighteningly easy: mmap something in `/sys/bus/pci/devices/*/resource*`.

For reference, you can use the pcie driver from litex: `https://github.com/enjoy-digital/litex/blob/master/litex/tools/remote/comm_pcie.py`

To actually implement it, create a new `BridgeKind` and `BridgeCore`: https://github.com/litex-hub/wishbone-utils/blob/master/wishbone-tool/src/bridge/mod.rs#L16-L30 . Then, Rust will complain loudly about incomplete enums.  Fix those errors and then you'll have a working PCIe core.

**Comment by xobs** [comment]
On the topic of Windows support: It seems as though the only official way to read/write PCIe on Windows is to create a full, signed driver.  You can put Windows into "test" mode to load unsigned drivers, but that's definitely onerous.

Having said that, and I in no way endorse doing this, there's a project called KDU that collects vulnerable drivers and abstracts them away to give `/dev/mem`-like access to programs with administrator privileges: https://github.com/hfiref0x/KDU/blob/218d9f8ab456303865c8c3334e79087b92bd31a6/Source/Hamakaze/kduprov.cpp

So definitely, if you have the ability to add support for this under Linux then by all means please do.

**Comment by dkozel** [comment]
Hi @xobs. I still have that Novena DVT1 around.

I have a Linux system, an Aller PCIe board, and a working LiteX build with Gen2 x4 PCIe running. So far so good. I've never written a line of Rust though so will probably make a C++ based tool to get familiar with `mmap`ing. If the yakshaving gets bad enough I'll learn some Rust and add support. I'm happy to support efforts to add PCIe support on Linux in any way from testing to remote access.

I may reach the point of writing a Windows driver for LitePCIe and testing it on macOS since this project is intended as a universal reference design. That's some months away though.

**Comment by xobs** [comment]
It looks like the current most-popular crate is `memmap` with over 2 million users.  That's a good sign.

The way I'd approach this is:

1. Install Rust
2. Create a new package: `cargo new mmap-test`
3. Add `memmap = "0.7"` to `Cargo.toml`
4. Check that it builds: `cargo check`
5. Try the following:

```rust
use memmap::MmapOptions;
use std::fs::OpenOptions;

fn main() {
    // Specify a real path here
    let file = OpenOptions::new()
        .read(true)
        .write(true)
        .open("/sys/bus/pci/devices/0000:00:14.0/resource0")
        .expect("Couldn't open PCIe BAR");
    let mut mmap = unsafe {
        MmapOptions::new()
            .map_mut(&file)
            .expect("Couldn't mmap PCIe BAR")
    };
    #[allow(clippy::cast_ptr_alignment)]
    let memory_range = mmap.as_mut_ptr() as *mut u32;
    unsafe {
        println!(
            "First two words of the boot ROM: {:08x} {:08x}",
            memory_range.add(0).read_volatile(),
            memory_range.add(1).read_volatile()
        )
    };
}
```

Assuming your boot ROM is mapped to offset `0x00000000`, this will read the first two words.  From here, it should be relatively easy to extend it into a Bridge by wrapping `read_volatile` and `write_volatile` into `do_peek` and `do_poke`.

**Comment by mithro** [comment]
I would be interested in seeing this work on the NeTV2 board! https://github.com/timvideos/litex-buildenv/wiki/HowTo-HDMI2PCIe-on-NeTV2

**Comment by xobs** [comment]
v0.6.12 should support pcie now:

```sh
$ sudo ./target/debug/wishbone-tool --pcie-bar /sys/bus/pci/devices/0000\:02\:00.0/resource0 --csr-csv ~/pcie_sdi/csr.csv dna_id
Value at 00005800: 00448580
$
```

Please let me know if it works for you!

**Comment by dkozel** [comment]
This works for me. My LitePCIe configuration sets the PCIe base address to that of the CSR region so lookups using the csr.csv document don't work.
https://github.com/enjoy-digital/litepcie_aller_test/blob/51b055ba85370b44a1e181d4204f8674850ef091/aller.py#L129

Initial thought is that just having an `--bar-offset` argument or similar might be the easiest way of getting that to work. Possible values could be a number (0x8200000) or string ("csr") which is the name of the memory region to lookup in the csr.csv.

```
if --bar-offset is defined
  if numeric
    apply
  else if csr-csv defined
    if bar-offset is blank/default
      lookup "csr" in csr.csv and apply
    else
      lookup value of string(bar-offset) in csr.csv and apply
```

**Comment by xobs** [comment]
I just tagged v0.6.16, which adds `--csr-csv-offset`.  Does that work for you?

**Comment by dkozel** [comment]
```
wishbone-tool --pcie-bar /sys/bus/pci/devices/0000\:03\:00.0/resource0 --csr-csv ~/src/litepcie_aller_test/soc_pciesoc_aller/csr.csv --csr-csv-offset 0x82000000 ctrl_scratch
thread 'main' panicked at 'attempt to add with overflow', src/config.rs:351:41
```
`csr_register,ctrl_scratch,0x82000004,1,rw`

I'm nearly positive that [wishbone-tool/src/config.rs#L351](https://github.com/litex-hub/wishbone-utils/blob/master/wishbone-tool/src/config.rs#L351) should be a subtraction. The offset is defined as a `u32` so cannot be negative.

[wishbone-tool/src/config.rs#L378](https://github.com/litex-hub/wishbone-utils/blob/master/wishbone-tool/src/config.rs#L378) probably also should be, but I get an `attempt to subtract with overflow` when I change them. I'll debug this but wanted to report back.

**Comment by xobs** [comment]
You're right, it's very broken.

I've reversed the order of operations and made some changes to how the offset works.  For starters, the argument is now `--register-offset` (however `--csr-csv-offset` will continue to work as an alias).  Additionally, the offset is now applied to ALL register accesses,  So for example if you do `wishbone-tool --register-offset 0x1000 0` it will refused to run, because address `0` is outside of the mappable range.  If you run `wishbone-tool --register-offset 0x1000 0x1000` it will work, and will read from Wishbone address 0x1000, which is called `0x0` in the bridge.

If this works for you, I'll create a new release.

**Comment by dkozel** [comment]
This works correctly. Tested single register reads and writes with both aliases. Tested the terminal server.

Very nice and thanks!

> (however --csr-csv-offset will continue to work as an alias). 

I still think being able to specify the name of a memory region to use as the offset would be user friendly. In the instantiation of the WishboneBridge the base_address is specified using the name of the memory region, the numerical value is an implementation detail not obvious to the user.

```
# Wishbone bridge
self.submodules.pcie_bridge = LitePCIeWishboneBridge(self.pcie_endpoint,
    base_address = self.mem_map["csr"])
```

The function `parse_csr_csv` returns a `HashMap<String, Option<u32>>`. If no `--register-offset` is supplied then the returned map will have the original addresses intact.

https://github.com/litex-hub/wishbone-utils/blob/a8f7a8f28209078aaa7c58abfc42c72f7caaa3ef/wishbone-tool/src/config.rs#L257

After this line there could be a bit of code to read the memory region address and loop over applying the offset (and it looks like discarding any elements with negative addresses).

I was going to try putting some Rust code in, but know the idoms well enough yet.

**Comment by xobs** [comment]
As you've seen, if you try to assign a negative value to a uint32 it will panic.  That's why I have it return an `Option<u32>`: `None` means that the mapping exists, but that it's out of range.

You're saying you think that `parse_csr_csv` should do it in two passes?  If it sees an offset for PCIe, then it should apply that offset if none exists?

What's the name of the PCIe region?  And does it always have the same name?  What does it look like in the csr.csv file?

**Comment by dkozel** [comment]
In my design, the PCIe Bar base address is assigned the value of `mem_map["csr"]`

In csr.csv that region is described as:
> memory_region,csr,0x82000000,65536,io

I suspect that this will be common to many designs, but there's no guarantee of that which is why I suggested specifying it by name. If someone were to set the `base_address` to a numerical value then the `--register-offset` would be entirely reasonable.

**Comment by dkozel** [comment]
One pass or two is an implementation detail. It seemed simple to leave the handling of the numerical offset as is since it's already done.

In C++ (ish) pseudo code:
```
offset = (--register-offset) ? --register-offset : register_mapping[--csr-csv-offset]
for k,v in register_mapping:
    register_mapping[k] = (v < offset) ? Null : v - offset
```

**Comment by xobs** [comment]
Can you see if that works for you?

**Comment by dkozel** [comment]
Its a thing of beauty! I'm all out of suggestions now. Thanks a ton, I'll make sure these features are put to use.

**Comment by xobs** [comment]
Released in v0.6.17
